doc/man/man3/ConfigurationFileReader.3
.TH "ConfigurationFileReader" 3 "Fri Mar 23 2018" "Version 1.0" "A Kalman Library (AKL)" \" -*- nroff -*-
.ad l
.nh
.SH NAME
ConfigurationFileReader
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <ConfigurationFileReader\&.h>\fP
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBFILTER_TYPES\fP { \fBMODEL_ROUKF\fP = 0, \fBMODEL_MAPPED_ROUKF\fP = 1 }"
.br
.ti -1c
.RI "enum \fBMAPPER_TYPES\fP { \fBIDENTITY\fP = 0, \fBEXPONENTIAL\fP = 1, \fBSIGMOIDAL\fP = 2 }"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBConfigurationFileReader\fP (string \fBfilename\fP)"
.br
.ti -1c
.RI "\fBAbstractROUKF\fP * \fBgetInstance\fP ()"
.br
.ti -1c
.RI "vector< double > \fBgetObservations\fP ()"
.br
.ti -1c
.RI "int \fBgetNParameters\fP () const"
.br
.ti -1c
.RI "int \fBgetNStates\fP () const"
.br
.ti -1c
.RI "int \fBgetNObservations\fP () const"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "string \fBfilename\fP"
.br
.ti -1c
.RI "vector< double > \fBobservations\fP"
.br
.ti -1c
.RI "int \fBnStates\fP"
.br
.ti -1c
.RI "int \fBnParameters\fP"
.br
.ti -1c
.RI "int \fBnObservations\fP"
.br
.ti -1c
.RI "\fBAbstractROUKF\fP * \fBroukfModel\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
File reader class that after parsing the target file generates a Kalman filter instance with appropriate mappers\&. 
.PP
Definition at line 20 of file ConfigurationFileReader\&.h\&.
.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBConfigurationFileReader::FILTER_TYPES\fP"
Types of filter that can be explicit in the configuration file\&. 
.PP
Definition at line 37 of file ConfigurationFileReader\&.h\&.
.SS "enum \fBConfigurationFileReader::MAPPER_TYPES\fP"
Types of mappers that can be explicit in the configuration file\&. 
.PP
Definition at line 39 of file ConfigurationFileReader\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "ConfigurationFileReader::ConfigurationFileReader (string filename)"
Reader constructor\&. 
.PP
Definition at line 24 of file ConfigurationFileReader\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "\fBAbstractROUKF\fP * ConfigurationFileReader::getInstance ()"
Returns the kalman instance associated with the configuration file\&. 
.PP
\fBReturns:\fP
.RS 4
Kalman model\&. 
.RE
.PP

.PP
Definition at line 32 of file ConfigurationFileReader\&.cpp\&.
.SS "int ConfigurationFileReader::getNObservations () const"
Getter for \fCnObservations\fP 
.PP
\fBReturns:\fP
.RS 4
\fCnObservations\fP 
.RE
.PP

.PP
Definition at line 182 of file ConfigurationFileReader\&.cpp\&.
.SS "int ConfigurationFileReader::getNParameters () const"
Getter for \fCnParameters\fP 
.PP
\fBReturns:\fP
.RS 4
\fCnParameters\fP 
.RE
.PP

.PP
Definition at line 172 of file ConfigurationFileReader\&.cpp\&.
.SS "int ConfigurationFileReader::getNStates () const"
Getter for \fCnStates\fP 
.PP
\fBReturns:\fP
.RS 4
\fCnStates\fP 
.RE
.PP

.PP
Definition at line 177 of file ConfigurationFileReader\&.cpp\&.
.SS "vector< double > ConfigurationFileReader::getObservations ()"
Returns the observations readed from the configuration file after executing getInstance\&. 
.PP
\fBReturns:\fP
.RS 4
Observations vector\&. 
.RE
.PP

.PP
Definition at line 168 of file ConfigurationFileReader\&.cpp\&.
.SH "Member Data Documentation"
.PP 
.SS "string ConfigurationFileReader::filename\fC [private]\fP"
Configuration file path\&. 
.PP
Definition at line 22 of file ConfigurationFileReader\&.h\&.
.SS "int ConfigurationFileReader::nObservations\fC [private]\fP"
Quantity of observations\&. 
.PP
Definition at line 30 of file ConfigurationFileReader\&.h\&.
.SS "int ConfigurationFileReader::nParameters\fC [private]\fP"
Quantity of parameters\&. 
.PP
Definition at line 28 of file ConfigurationFileReader\&.h\&.
.SS "int ConfigurationFileReader::nStates\fC [private]\fP"
Quantity of internal states\&. 
.PP
Definition at line 26 of file ConfigurationFileReader\&.h\&.
.SS "vector<double> ConfigurationFileReader::observations\fC [private]\fP"
Observations loaded from configuration file 
.PP
Definition at line 24 of file ConfigurationFileReader\&.h\&.
.SS "\fBAbstractROUKF\fP* ConfigurationFileReader::roukfModel\fC [private]\fP"
Singleton attribute of the generated kalman filter\&. 
.PP
Definition at line 32 of file ConfigurationFileReader\&.h\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for A Kalman Library (AKL) from the source code\&.


io/ConfigurationFileReader.cpp
/*
 * ConfigurationReader.cpp
 *
 *  Created on: Feb 27, 2018
 *      Author: Gonzalo D. Maso Talou
 */

#include "ConfigurationFileReader.h"

#include "../MappedROUKF.h"
#include "../ROUKF.h"

#include "../SigmaPointsGenerator.h"

#include "../mapping/AbstractParameterMapper.h"
#include "../mapping/IdentityParameterMapper.h"
#include "../mapping/ExponentialParameterMapper.h"
#include "../mapping/SigmoidParameterMapper.h"

#include <libconfig.h++>

using namespace libconfig;

ConfigurationFileReader::ConfigurationFileReader(string filename) {
	this->filename = filename;
	this->nStates = 0;
	this->nParameters = 0;
	this->nObservations = 0;
	this->roukfModel = NULL;
}

AbstractROUKF* ConfigurationFileReader::getInstance() {

	if(roukfModel)
		return roukfModel;

	Config config;

	try {
		config.readFile(filename.c_str());
	} catch (const FileIOException &fioex) {
		std::cerr << "I/O error while reading configuration file for ROUKF." << std::endl;
    } catch (const ParseException &pex) {
        std::cerr << "Parse error at " << pex.getFile() << ":" << pex.getLine()
                << " - " << pex.getError() << std::endl;
    }

	int typeROUKF = 0;
	try {
		typeROUKF = config.lookup("FilterType");
	} catch (const SettingNotFoundException &nfex) {
		cerr << "Configuration file must explicit the FilterType (\"MappedROUKF\" | \"ROUKF\")." << endl;
	}

	try {
		nStates = config.lookup("States");
		nParameters = config.lookup("Parameters");
		nObservations = config.lookup("Observations");
	} catch (const SettingNotFoundException &nfex) {
		cerr << "States, Parameters or Observations number is missing." << endl;
	}

	vector<double> initialGuess;
	vector<double> parameterUncertainty;
	try {
		Setting& sVectorDoubles = config.lookup("InitialGuess");
		for (int i = 0; i < sVectorDoubles.getLength(); ++i) {
			initialGuess.push_back((double) sVectorDoubles[i]);
		}
		Setting& sVectorDoubles2 = config.lookup("ParameterUncertainty");
		for (int i = 0; i < sVectorDoubles2.getLength(); ++i) {
			parameterUncertainty.push_back((double) sVectorDoubles2[i]);
		}
	} catch (const SettingNotFoundException &nfex) {
		cerr << "Error while reading InitialGuess or ParameterUncertainty fields." << endl;
	}

	vector<AbstractParameterMapper *> parameterMapping;
	vector<int> parametersPerMapping;
	try {
		Setting& mapperList = config.lookup("ParameterMapping");
		for (int i = 0; i < mapperList.getLength(); ++i) {
			int mapType = (int) mapperList[i]["type"];
			switch (mapType) {
			case IDENTITY:
				parametersPerMapping.push_back(mapperList[i]["numParam"]);
				parameterMapping.push_back(new IdentityParameterMapper());
				break;
			case EXPONENTIAL:
				parametersPerMapping.push_back((int) mapperList[i]["numParam"]);
				parameterMapping.push_back(new ExponentialParameterMapper());
				break;
			case SIGMOIDAL:
				parametersPerMapping.push_back(mapperList[i]["numParam"]);
				parameterMapping.push_back(new SigmoidParameterMapper((double) (mapperList[i]["min"]), (double) (mapperList[i]["max"])));
				break;
			default:
				parametersPerMapping.push_back(mapperList[i]["numParam"]);
				parameterMapping.push_back(new IdentityParameterMapper());
				break;
			}
		}
	} catch (const SettingNotFoundException &nfex) {
		cerr << "Error while reading ParameterMapping fields." << endl;
	}

	vector<double> observationsUncertainty;
	try {
		Setting& sVectorDoubles = config.lookup("ObservationsValues");
		for (int i = 0; i < sVectorDoubles.getLength(); ++i) {
			observations.push_back((double) sVectorDoubles[i]);
		}
		Setting& sVectorDoubles2 = config.lookup("ObservationsUncertainty");
		for (int i = 0; i < sVectorDoubles2.getLength(); ++i) {
			observationsUncertainty.push_back((double) sVectorDoubles2[i]);
		}
	} catch (const SettingNotFoundException &nfex) {
		cerr << "Error while reading ObservationsValues or ObservationUncertainty fields." << endl;
	}

	SigmaPointsGenerator::SIGMA_DISTRIBUTION sigmaDistribution = SigmaPointsGenerator::SIMPLEX;
	try {
		int sigmaType = config.lookup("SigmaDistribution");
		sigmaDistribution = static_cast<SigmaPointsGenerator::SIGMA_DISTRIBUTION>(sigmaType);
	} catch (const SettingNotFoundException &nfex) {
		cerr << "Error while reading SigmaDistribution field." << endl;
	}

	double tol = 1E-5;
	try {
		tol = config.lookup("ConvergenceTol");
	} catch (const SettingNotFoundException &nfex) {
		cerr << "Error while reading ConvergenceTol field." << endl;
	}

	int maxIt = 1000;
	try {
		maxIt = config.lookup("MaxIterations");
	} catch (const SettingNotFoundException &nfex) {
		cerr << "Error while reading MaxIterations field." << endl;
	}

	switch (typeROUKF) {
	case MODEL_ROUKF:
		roukfModel = new ROUKF(nObservations, nStates, nParameters,
				&(observationsUncertainty[0]), &(parameterUncertainty[0]),
				sigmaDistribution);
		break;
	case MODEL_MAPPED_ROUKF:
		roukfModel = new MappedROUKF(nObservations, nStates, nParameters,
				observationsUncertainty, parameterUncertainty, sigmaDistribution,
				new CompositeParameterMapper(parametersPerMapping, parameterMapping));
		break;
	default:
		roukfModel = new ROUKF(nObservations, nStates, nParameters,
				&(observationsUncertainty[0]), &(parameterUncertainty[0]),
				sigmaDistribution);
		break;
	}
	roukfModel->setParameters(&(initialGuess[0]));
	roukfModel->setTolerance(tol);
	roukfModel->setMaxIterations(maxIt);

	return roukfModel;

}

vector<double> ConfigurationFileReader::getObservations() {
	return observations;
}

int ConfigurationFileReader::getNParameters() const
{
	return nParameters;
}

int ConfigurationFileReader::getNStates() const
{
	return nStates;
}

int ConfigurationFileReader::getNObservations() const
{
	return nObservations;
}
